File: src/capsolver/browser/launcher.py
"""Browser launcher using zendriver (nodriver successor) — undetected, headless=False."""

from __future__ import annotations

import shutil
import uuid
from pathlib import Path

import zendriver as zd
from zendriver.core.config import Config

from capsolver.browser.page_adapter import PageAdapter
from capsolver.browser.profile import extension_ids_for, extensions_to_load, warmup_profile
from capsolver.browser.profile_lock import clean_stale_locks, profile_file_lock
from capsolver.core.config import AppConfig, get_config
from capsolver.core.logging import get_logger
from capsolver.core.platform import PlatformInfo, detect_platform, ensure_display

logger = get_logger(__name__)

BROWSER_BINARIES = ("brave", "brave-browser", "google-chrome-stable", "google-chrome", "chromium")

BLOCKED_BROWSER_ARGS = frozenset({
    "--no-sandbox",
    "--disable-setuid-sandbox",
    "--enable-automation",
    "--remote-debugging-port",
})


def find_browser_binary(config: AppConfig) -> str | None:
    if config.browser.executable_path:
        p = Path(config.browser.executable_path)
        if p.exists():
            return str(p)
    for name in BROWSER_BINARIES:
        found = shutil.which(name)
        if found:
            return found
    return None


def _extra_browser_args(config: AppConfig) -> list[str]:
    args: list[str] = []
    for arg in config.browser.args:
        base = arg.split("=", 1)[0]
        if base in BLOCKED_BROWSER_ARGS:
            logger.warning("blocked_browser_arg", arg=arg)
            continue
        args.append(arg)
    return args


def _build_zendriver_config(
    config: AppConfig,
    profile_dir: Path,
    extension_paths: list[Path],
    binary: str | None,
) -> Config:
    browser_type = "brave" if binary and "brave" in binary.lower() else "auto"
    zc = Config(
        user_data_dir=str(profile_dir),
        headless=False,
        browser_executable_path=binary,
        browser=browser_type,  # type: ignore[arg-type]
        sandbox=True,
        browser_args=_extra_browser_args(config),
    )
    if extension_paths:
        for path in extension_paths:
            logger.info("extension_registered", path=str(path.resolve()))
        joined = ",".join(str(p.resolve()) for p in extension_paths)
        zc.add_argument(f"--disable-extensions-except={joined}")
        zc.add_argument(f"--load-extension={joined}")
        # Chrome/Brave 127+ ignore --load-extension unless this feature stays disabled.
        # zendriver appends its own --disable-features (without this flag), which would
        # override ours, so this must be the LAST --disable-features on the command line.
        zc.add_argument(
            "--disable-features=IsolateOrigins,site-per-process,"
            "DisableLoadExtensionCommandLineSwitch"
        )
    return zc


class BrowserHandle:
    """Zendriver browser + active tab."""

    def __init__(self, browser: zd.Browser, tab: zd.Tab, page: PageAdapter):
        self.browser = browser
        self.tab = tab
        self.page = page

    @property
    def pages(self) -> list[PageAdapter]:
        return [self.page]


class BrowserLauncher:
    """
    Launch Brave/Chrome via zendriver.

    Extensions load at browser start when load_extensions_at_startup=true.
    """

    def __init__(self, config: AppConfig | None = None, platform_info: PlatformInfo | None = None):
        self.config = config or get_config()
        self.platform = platform_info or detect_platform()

    async def start(self) -> None:
        if not self.platform.has_display and self.platform.xvfb_available:
            ensure_display(self.platform, self.config.platform.display)
        logger.info("zendriver_ready", platform=self.platform.os_type.value)

    async def stop(self) -> None:
        pass

    async def release_shared_context(self, session_id: str) -> None:
        pass

    def _profile_dir(self, session_id: str) -> Path:
        cfg = self.config.browser
        if cfg.system_profile_path:
            p = Path(cfg.system_profile_path).expanduser()
            if p.exists():
                return p.resolve()
        if cfg.shared_profile:
            return self.config.resolve_path(cfg.user_data_base) / f"shared_{cfg.shared_profile_id}"
        return self.config.resolve_path(cfg.user_data_base) / "temp" / session_id

    def _should_include_extensions(self, with_extensions: bool | None) -> bool:
        if with_extensions is not None:
            return with_extensions
        return self.config.browser.load_extensions_at_startup

    async def create_session(
        self,
        session_id: str | None = None,
        with_extensions: bool | None = None,
        *,
        configure_nopecha: bool = True,
    ) -> tuple[BrowserHandle, str, dict[str, str]]:
        await self.start()

        include_ext = self._should_include_extensions(with_extensions)
        extension_paths = await extensions_to_load(self.config, include=include_ext)
        sid = session_id or str(uuid.uuid4())
        profile_dir = self._profile_dir(sid)
        profile_dir.mkdir(parents=True, exist_ok=True)
        clean_stale_locks(profile_dir)

        binary = find_browser_binary(self.config)
        zc = _build_zendriver_config(self.config, profile_dir, extension_paths, binary)

        with profile_file_lock(profile_dir):
            logger.info(
                "launching_zendriver",
                session_id=sid,
                profile=str(profile_dir),
                headless=False,
                sandbox=True,
                extensions=[p.name for p in extension_paths],
                binary=binary or zc.browser,
            )
            browser = await zd.Browser.create(zc)

        tab = browser.main_tab
        if tab is None:
            tab = await browser.get("about:blank")
        page = PageAdapter(browser, tab)

        captcha = self.config.automation.poketwo.captcha
        extension_ids = await warmup_profile(
            browser,
            extension_paths,
            nopecha_api_key=captcha.nopecha_api_key,
            configure_nopecha=configure_nopecha and bool(extension_paths),
        )
        if not extension_ids and extension_paths:
            extension_ids = extension_ids_for(extension_paths)

        handle = BrowserHandle(browser, tab, page)
        return handle, sid, extension_ids

    async def cleanup_session(self, session_id: str) -> None:
        if self.config.browser.system_profile_path or self.config.browser.shared_profile:
            return
        for base in (
            self.config.resolve_path(self.config.browser.user_data_base) / "temp" / session_id,
            self.config.resolve_path(self.config.browser.user_data_base) / session_id,
        ):
            if base.exists():
                shutil.rmtree(base, ignore_errors=True)
                logger.info("temp_profile_cleaned", session_id=session_id, path=str(base))

File: src/browser/launcher.js
import { chromium } from "playwright";
import { v4 as uuidv4 } from "uuid";
import fs from "fs";
import path from "path";
import { getConfig, resolvePath } from "../config.js";
import { log } from "../logger.js";
import { detectPlatform, ensureDisplay } from "../platform.js";
import { extensionsToLoad, extensionIdsFor, findBrowserBinary } from "./extensions.js";
import { withProfileLock } from "./profile-lock.js";

const STEALTH_INIT_SCRIPT = `
Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
window.chrome = window.chrome || { runtime: {} };
`;

const BLOCKED_ARGS = new Set([
  "--no-sandbox",
  "--disable-setuid-sandbox",
  "--enable-automation",
  "--remote-debugging-port",
]);

export class PlaywrightLauncher {
  constructor(config = getConfig()) {
    this.config = config;
    this.platform = detectPlatform();
  }

  _profileDir(sessionId) {
    const cfg = this.config.browser;
    if (cfg.systemProfilePath && fs.existsSync(cfg.systemProfilePath)) {
      return path.resolve(cfg.systemProfilePath);
    }
    if (cfg.sharedProfile) {
      return resolvePath(path.join(cfg.userDataBase, `shared_${cfg.sharedProfileId}`));
    }
    return resolvePath(path.join(cfg.userDataBase, "temp", sessionId));
  }

  async createSession({ sessionId, withExtensions } = {}) {
    if (!this.platform.hasDisplay && this.platform.xvfbAvailable) {
      ensureDisplay(this.config.platform.display);
    }

    const includeExt =
      withExtensions !== undefined ? withExtensions : this.config.browser.loadExtensionsAtStartup;

    const extensionPaths = await extensionsToLoad({ include: includeExt });
    const sid = sessionId || uuidv4();
    const profileDir = this._profileDir(sid);
    fs.mkdirSync(profileDir, { recursive: true });

    const binary = findBrowserBinary();
    const args = ["--disable-blink-features=AutomationControlled"];
    if (extensionPaths.length) {
      const joined = extensionPaths.join(",");
      args.push(`--disable-extensions-except=${joined}`, `--load-extension=${joined}`);
    }
    for (const arg of this.config.browser.args) {
      const base = arg.split("=")[0];
      if (!BLOCKED_ARGS.has(base)) args.push(arg);
      else log("WARN", "blocked_browser_arg", "", { arg });
    }

    const launchOpts = {
      headless: false,
      executablePath: binary || undefined,
      args,
      ignoreDefaultArgs: ["--enable-automation"],
      viewport: {
        width: this.config.browser.viewportWidth,
        height: this.config.browser.viewportHeight,
      },
    };

    log("INFO", "launching_playwright", "", {
      sessionId: sid,
      profile: profileDir,
      extensions: extensionPaths.map((p) => path.basename(p)),
      binary,
    });

    const context = await withProfileLock(profileDir, async () =>
      chromium.launchPersistentContext(profileDir, launchOpts)
    );

    await context.addInitScript(STEALTH_INIT_SCRIPT);
    const page = context.pages()[0] || (await context.newPage());
    const extensionIds = extensionIdsFor(extensionPaths);

    const nopechaKey = this.config.automation.poketwo.captcha.nopechaApiKey;
    if (nopechaKey && extensionIds.nopecha) {
      await this._configureNopecha(page, nopechaKey);
    }

    log("INFO", "session_ready", "", { extensions: Object.keys(extensionIds) });
    return { context, page, profileDir, sessionId: sid, extensionIds };
  }

  async _configureNopecha(page, apiKey) {
    try {
      await page.goto(`https://nopecha.com/setup?api_key=${apiKey}`, {
        waitUntil: "domcontentloaded",
        timeout: 60000,
      });
      await page.waitForTimeout(3000);
      log("INFO", "nopecha_api_key_configured");
    } catch (e) {
      log("WARN", "nopecha_setup_failed", String(e));
    }
  }

  async cleanupSession(sessionId, profileDir) {
    const cfg = this.config.browser;
    if (cfg.systemProfilePath || cfg.sharedProfile) return;
    for (const base of [
      resolvePath(path.join(cfg.userDataBase, "temp", sessionId)),
      resolvePath(path.join(cfg.userDataBase, sessionId)),
      profileDir,
    ]) {
      if (base && fs.existsSync(base)) {
        fs.rmSync(base, { recursive: true, force: true });
        log("INFO", "temp_profile_deleted", "", { sessionId, path: base });
      }
    }
  }
}

export class BrowserSession {
  constructor({ launcher, sessionId, context, page, profileDir, extensionIds, artifactsDir }) {
    this.launcher = launcher;
    this.sessionId = sessionId;
    this.context = context;
    this.page = page;
    this.profileDir = profileDir;
    this.extensionIds = extensionIds;
    this.artifactsDir = artifactsDir;
    this._closed = false;
    fs.mkdirSync(artifactsDir, { recursive: true });
  }

  async screenshot(name) {
    const filePath = path.join(this.artifactsDir, `${name}.png`);
    try {
      await this.page.screenshot({ path: filePath, fullPage: true });
      log("INFO", "screenshot_captured", "", { sessionId: this.sessionId, path: filePath });
    } catch (e) {
      log("WARN", "screenshot_skipped", String(e), { sessionId: this.sessionId });
    }
    return filePath;
  }

  async close(cleanupProfile = true) {
    if (this._closed) return;
    this._closed = true;
    try {
      await this.context.close();
    } catch (e) {
      log("WARN", "browser_stop_error", String(e), { sessionId: this.sessionId });
    }
    if (cleanupProfile) {
      await this.launcher.cleanupSession(this.sessionId, this.profileDir);
    }
    log("INFO", "session_closed", "", { sessionId: this.sessionId });
  }
}

export class BrowserPool {
  constructor(config = getConfig()) {
    this.config = config;
    this.launcher = new PlaywrightLauncher(config);
    this.maxConcurrent = config.browser.maxConcurrent;
    this._active = new Map();
    this._queue = [];
    this._running = 0;
  }

  get activeCount() {
    return this._active.size;
  }

  async acquire(jobId) {
    await new Promise((resolve) => {
      const tryAcquire = () => {
        if (this._running < this.maxConcurrent) {
          this._running++;
          resolve();
        } else {
          this._queue.push(tryAcquire);
        }
      };
      tryAcquire();
    });

    const artifactsDir = path.join(resolvePath(this.config.jobs.artifactsDir), jobId);
    const suffix = uuidv4().slice(0, 8);
    const { context, page, profileDir, sessionId, extensionIds } =
      await this.launcher.createSession({
        sessionId: `${jobId}_${suffix}`,
        withExtensions: this.config.browser.loadExtensionsAtStartup,
      });

    const session = new BrowserSession({
      launcher: this.launcher,
      sessionId,
      context,
      page,
      profileDir,
      extensionIds,
      artifactsDir,
    });
    this._active.set(sessionId, session);
    log("INFO", "session_acquired", "", { jobId, sessionId, active: this.activeCount });
    return session;
  }

  async release(session, cleanupProfile = true) {
    const sessionId = session.sessionId;
    await session.close(cleanupProfile);
    this._active.delete(sessionId);
    this._running--;
    if (this._queue.length) {
      const next = this._queue.shift();
      next();
    }
    log("INFO", "session_released", "", { sessionId, active: this.activeCount });
  }

  async shutdown() {
    for (const session of [...this._active.values()]) {
      await this.release(session);
    }
  }
}
